lib/parallel_guest_migration_base.pm

```diff
@@ -1255,58 +1345,187 @@ EOF
-                    $_temp |= 1 && record_info("Network $guest_matrix{$_guest}{netname} creation failed", script_output("ip addr show all;ip route show all", type_command => 1, proceed_on_failure => 1), result => 'fail') if (!$_find_netdev);
-                }
-                else {
-                    $_temp = 0;
-                }
-                my $_dnsmasq_command = "/usr/sbin/dnsmasq --bind-dynamic --listen-address=$_ipaddr --dhcp-range=$_startaddr,$_endaddr,$_netmask,8h --interface=br123 --dhcp-authoritative --no-negcache --dhcp-option=option:router,$_ipaddr --log-queries --log-dhcp --dhcp-sequential-ip --dhcp-client-update --no-daemon";
-                if (!script_output("ps ax | grep -i \"$_dnsmasq_command\" | grep -v grep | awk \'{print \$1}\'", type_command => 1, proceed_on_failure => 1)) {
-                    $_temp |= script_run("((nohup $_dnsmasq_command) &)");
-                    my $_find_dnsmasq = script_output("ps ax | grep -i \"$_dnsmasq_command\" | grep -v grep | awk \'{print \$1}\'", type_command => 1, proceed_on_failure => 1);
-                    $_temp |= 1 && record_info("DHCP service failed on $guest_matrix{$_guest}{netname}", "Command to start DHCP service is $_dnsmasq_command", result => 'fail') if (!$_find_dnsmasq);
-                }
+            if (($_temp == 0) and ($args{_usedns} == 1) and ($_guest_matrix{$_guest}{netname} ne 'default')) {
+                $_temp |= config_domain_resolver(resolvip => $_ipaddr, domainname => $_guest_params{dns_domainname});
             }
         }
-        push(@_guest_network_configured, $guest_matrix{$_guest}{netname});
+        elsif ($_guest_matrix{$_guest}{nettype} eq 'bridge') {
+            $_temp = $self->create_guest_network_bridge_device(_guest => $_guest, _ipaddr => $_ipaddr, _masklen => $_masklen);
+            $_temp |= $self->create_guest_network_bridge_service(_guest => $_guest, _ipaddr => $_ipaddr, _netmask => $_netmask, _masklen => $_masklen, _startaddr => $_startaddr, _endaddr => $_endaddr) if ($_temp == 0);
+        }
+        push(@_guest_network_configured, $_guest_matrix{$_guest}{netname});
         $_ret |= $_temp;
         save_screenshot;
     }
     record_info("Guest network configuration done", script_output("ip addr show;ip route show all;virsh $_uri net-list --all;(for i in \`virsh $_uri net-list --all --name\`;do virsh $_uri net-dumpxml \$i;done);ps axu | grep dnsmasq", type_command => 1, proceed_on_failure => 1));
-    croak("Guest network creation failed for certain guest") if ($_ret != 0 and $args{die} == 1);
+    croak("Guest network creation failed for certain guest") if ($_ret != 0 and $args{_die} == 1);
+    return $_ret;
+}
+
+
+=head2 create_guest_network_bridge_device
+
+Create network bridge device br0 (host bridge) or brXXX (internal bridge) by
+using derived information from host or in data structure %_guest_matrix. This
+should be called in create_guest_network.
+=cut
+
+sub create_guest_network_bridge_device {
```

Comment:
This is another copy of bridge setup, in addition to the one in GI  and another one in host installation(although not stable yet). I understand there are differences...
I have the feeling that we have given ourselves too much network related code maintenance work here and there. Is it possible to think about reuse or just let the bridge setup done outside of migration? We do not need to let every function test be comprehensive.
@waynechen55  @Julie-CAO HDYT?

Reply:
> This is another copy of bridge setup, in addition to the one in GI and another one in host installation(although not stable yet). I understand there are differences... I have the feeling that we have given ourselves too much network related code maintenance work here and there. Is it possible to think about reuse or just let the bridge setup done outside of migration? We do not need to let every function test be comprehensive. @waynechen55 @Julie-CAO HDYT?

I have always preferred things to be simpler, to leave the host to install br0. but if for workaround or debug purpose since br0 is not available for now, I can understand.

Reply:
I refactored further to move network device part into ```lib/virt_autotest/virtual_network_utils.pm```, so no any duplication and they become common functions can be used and called conveniently:
```config_domain_resolver```
```write_network_bridge_device_config```
```write_network_bridge_device_ifcfg```
```write_network_bridge_device_nmconnection```
```ctivate_network_bridge_device```
```config_virtual_network_device```

Reply:
Cool!